src/coreclr/jit/gentree.cpp

```diff
@@ -23897,6 +23896,35 @@ GenTree* Compiler::gtNewSimdSumNode(var_types type, GenTree* op1, CorInfoType si
         intrinsic = NI_SSSE3_HorizontalAdd;
     }
 
+    if (simdSize == 64)
+    {
+        assert(IsBaselineVector512IsaSupportedDebugOnly());
+        // This is roughly the following managed code:
+        //   ...
+        //   simd64 tmp2 = tmp1;
+        //          tmp3 = tmp2.GetUpper();
+        //   simd32 tmp4 = Isa.Add(tmp1.GetLower(), tmp2);
+        //          tmp5 = tmp4;
+        //   simd16 tmp6 = tmp4.GetUpper();
+        //       tmp1 = Isa.Add(tmp1.GetLower(), tmp2);
+        //   ...
+        // From here on we can treat this as a simd16 reduction
+        GenTree* op1Dup     = fgMakeMultiUse(&op1);
+        GenTree* op1Lower32 = gtNewSimdGetUpperNode(TYP_SIMD32, op1Dup, simdBaseJitType, simdSize);
+        GenTree* op1Upper32 = gtNewSimdGetLowerNode(TYP_SIMD32, op1, simdBaseJitType, simdSize);
+
+        simdSize   = simdSize / 2;
+        op1Lower32 = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD32, op1Lower32, op1Upper32, simdBaseJitType, simdSize);
+        haddCount--;
+
+        GenTree* op1Dup32   = fgMakeMultiUse(&op1Lower32);
+        GenTree* op1Lower16 = gtNewSimdGetUpperNode(TYP_SIMD16, op1Lower32, simdBaseJitType, simdSize);
+        GenTree* op1Upper16 = gtNewSimdGetLowerNode(TYP_SIMD16, op1Dup32, simdBaseJitType, simdSize);
+        simdSize            = simdSize / 2;
+        op1                 = gtNewSimdBinOpNode(GT_ADD, TYP_SIMD16, op1Lower16, op1Upper16, simdBaseJitType, simdSize);
+        haddCount--;
+    }
```

Comment:
This is "correct", but has an interesting side effect in that can change the result for float/double.

Since float/double are not associative due to the rounding, summing accross as `[0] + [1] + [2] + [3] + ...` is different from summing pairwise as `(([0] + [1]) + ([2] + [3])) + ...`, which is different from summing per lane, then combining the lanes, etc.

Today, we're basically doing it per lane, then combining lanes. Within that lane we're typically doing `pairwise` because that's how `addv` (add across) works on Arm64, it's how `hadd` (horizontal add) works on x86/x64, and its trivial to emulate using `shufps` on older hardware.

We don't really want to have the results subtly change based on what the hardware supports, so we should probably try to ensure this keeps things overall consistent in how it operates.

Reply:
I'm busy with something else but will get back to this once I'm done. Sorry for the delay

Reply:
No worries. This one isn't "critical" for .NET 8 and we're already generating decent (but not amazing) code that would be similar to what a user might manually write.